### PR TITLE
Hide superseded rooms from the room list using dynamic room predecessors

### PR DESCRIPTION
**Pull request by andybalaam**
Pass the dynamic predecessor feature flag when listing rooms

After this change, rooms that are superseded by another room using the MSC3946 mechanism won't be visible in the room list.

Towards [MSC3946](https://github.com/matrix-org/matrix-spec-proposals/pull/3946)

Part of https://github.com/vector-im/element-web/issues/24417

rec-2023-02-03_13 24 02

---
Here's what your changelog entry will look like:

## ✨ Features
 * Hide superseded rooms from the room list using dynamic room predecessors ([\#10068](https://github.com/matrix-org/matrix-react-sdk/pull/10068)). Contributed by @andybalaam.

### CODE REVIEW

File: src/stores/room-list/RoomListStore.ts [hunk 1]
```diff
@@ -496,7 +496,8 @@ export class RoomListStoreClass extends AsyncStoreWithClient<IState> implements
     private getPlausibleRooms(): Room[] {
         if (!this.matrixClient) return [];
 
-        let rooms = this.matrixClient.getVisibleRooms().filter((r) => VisibilityProvider.instance.isRoomVisible(r));
+        let rooms = this.matrixClient.getVisibleRooms(SettingsStore.getValue("feature_dynamic_room_predecessors"));
```
Comment (t3chguy) [inline src/stores/room-list/RoomListStore.ts:499]: We should avoid SettingsStore calls on the critical codepaths such as this as they end up being expensive

Comment (andybalaam) [reply src/stores/room-list/RoomListStore.ts:499]: Where should we call it?

Comment (t3chguy) [reply src/stores/room-list/RoomListStore.ts:499]: Class constructor and set up a watcher/monitor to respond to updates and recalculate the room list when such an update happens

**Comment by andybalaam** [comment]
I will address the strict error in a separate change.

**Comment by andybalaam** [comment]
SonarCloud wants me to test all functionality of `RoomListStore` which I argue is out of scope of this change.

**Pull request merged**